# Solutions to Project Euler Problems
This notebook is a collection of my solutions for the Project Euler problems (found here: https://projecteuler.net/) done using Python code.
Answering these problems aims to provide me with the following:

* A gauge of my skills in functional programming/problem-solving, at the moment
* A space to help me see errors, redundancies, and bad practices I do with my code
* An opportunity to help me learn more about mathematical problems and how to solve them using code
* An eventual collection that I could add to my portfolio so potential employers could see my thought process when solving problems

## **1. Multiples of 3 and 5**
If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [1]:
# Create a function which takes in a number, then returns the sum of all multiples of 3 or 5 below that number 
def sum_of_multiples(number):
    multiples = []
    for n in range(number):
        if (n % 3 == 0) | (n % 5 == 0):
            multiples.append(n)
    return print(sum(multiples))

sum_of_multiples(1000)

233168


### Answer: 233168

## **2. Even Fibonacci Numbers**
Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

`1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...`

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [2]:
terms = [1, 2]
even_terms = [2]
    
while max(terms) < 4000000:
    first = terms[-2]                      # The first number will always be the second-to-the-last number in the 'terms' list
    second = terms[-1]                     # The second number will always be the latest number to be added into the 'terms' list
    next_term = first + second
    if next_term % 2 == 0:
        even_terms.append(next_term)       # Appends the term if it is an even number
    terms.append(next_term)

print(sum(even_terms))

4613732


### Answer: 4613732

## **3. Largest Prime Factor**
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?


### Solution

In order to get the prime factors of 13195, a mathematical operation called *prime factorization* was used. A simple way to do this is to divide the number by each prime number in ascending order. 

* **If the answer is not a whole number:** we try again using the next prime number.
* **If the answer is a whole number:** we get the quotient and then divide that by the next prime number in the list. We end our operation when we reach a quotient of 1.

### Example:

In [3]:
# To get the prime factors of 13195, we start by dividing it by the smallest prime number (except 1), then working our way up the list
13195 / 2 # Answer is not a whole number
13195 / 3 # Answer is not a whole number
13195 / 5 # Answer is 2639
2639 / 7 # Answer is 377
377 / 11 # Answer is not a whole number
377 / 13 # Answer is 29
29 / 29 # Answer is 1. We end here

1.0

Thus, the prime factors of 13195 are: 5, 7, 13 and 29.

Now let's make a program that would find the prime factors of any number we want.

In [ ]:
def get_prime_factors(number):
    for i in range(2, number + 1):
        if(number % i == 0):
            isprime = 1
            for j in range(2, (i //2 + 1)):
                if(i % j == 0):
                    isprime = 0
                    break      
            if (isprime == 1):
                print(" %d is a prime factor of %d" %(i, number))
            
get_prime_factors(600851475143)

 71 is a prime factor of 600851475143
 839 is a prime factor of 600851475143
 1471 is a prime factor of 600851475143
 6857 is a prime factor of 600851475143


### Answer: The prime factors of 600851475143 are:
* 71
* 839
* 1471
* 6857

### with 6857 being the largest.

## **4. Largest Palindrome Product**
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.


## **5. Smallest Multiple**
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

### Solution

The simplest way to answer this is by using the numpy.lcm.reduce() method and use the range of numbers as arguments.

In [ ]:
import pandas as pd
import numpy as np

foo = list(range(1, 21))

ans = np.lcm.reduce(foo)
print(" The smallest multiple of", foo[0], "to", foo[-1], "is", ans)

To build a solution manually, we could use our prime factorization function from before. One way of getting the LCM of a list of numbers is multiplying the highest power of each prime number together (meaning the last 2 prime factors for each number).

In [ ]:
# To be continued